onnxruntime/core/platform/windows/device_discovery.cc

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/platform/device_discovery.h"
+
+#include <unordered_set>
+
+#include <initguid.h>
+#include <dxcore.h>
+#include <dxcore_interface.h>
+#include <wil/com.h>
+
+// GetDxgiInfo
+#include <dxgi.h>
+#pragma comment(lib, "dxgi.lib")
+#include <iostream>
+
+#include "core/common/cpuid_info.h"
+#include "core/session/abi_devices.h"
+
+// TEMPORARY: The CI builds target Windows 10 so do not have these GUIDs.
+// This is to make the builds pass so any other issues can be resolved, but needs a real solution prior to checkin.
+// these values were added in 10.0.22621.0 as part of DirectXCore API
+#if NTDDI_VERSION < NTDDI_WIN10_RS5
+DEFINE_GUID(DXCORE_ADAPTER_ATTRIBUTE_D3D12_GENERIC_ML, 0xb71b0d41, 0x1088, 0x422f, 0xa2, 0x7c, 0x2, 0x50, 0xb7, 0xd3, 0xa9, 0x88);
+DEFINE_GUID(DXCORE_HARDWARE_TYPE_ATTRIBUTE_NPU, 0xd46140c4, 0xadd7, 0x451b, 0x9e, 0x56, 0x6, 0xfe, 0x8c, 0x3b, 0x58, 0xed);
+#endif
+
+namespace onnxruntime {
+
+namespace {
+std::unordered_set<OrtHardwareDevice> GetInferencingDevices() {
+  std::unordered_set<OrtHardwareDevice> found_devices;
+
+  // Get information about the CPU device
+  auto vendor = CPUIDInfo::GetCPUIDInfo().GetCPUVendor();
+
+  // Create an OrtDevice for the CPU and add it to the list of found devices
+  OrtHardwareDevice cpu_device{OrtHardwareDeviceType_CPU, 0, std::string(vendor), 0};
+
+  found_devices.insert(cpu_device);
+
+  // Get all GPUs and NPUs by querying WDDM/MCDM.
+  wil::com_ptr<IDXCoreAdapterFactory> adapterFactory;
```

Comment:
The naming convention is inconsistent here, don't we use snake_case vs camelCase? (my guess is it's Windows code influcence)

Reply:
can cleanup. this was contributed as starter code. for this PR assume this is all replaced and we will make it conform to coding standards when the production ready implementation is added

Reply:
will cleanup in the 'part 2' changes rather than here.